src/workerd/api/worker-rpc.c++

```diff
@@ -62,50 +122,47 @@ kj::Promise<capnp::Response<rpc::JsRpcTarget::CallResults>> WorkerRpc::sendWorke
   }
 
   auto callResult = builder.send();
-  auto customEventResult = worker->customEvent(kj::mv(event)).attach(kj::mv(worker));
-
-  // If customEvent throws, we'll cancel callResult and propagate the exception. Otherwise, we'll
-  // just wait until callResult finishes.
-  return callResult.exclusiveJoin(customEventResult
-      .then([](auto&&) -> kj::Promise<capnp::Response<rpc::JsRpcTarget::CallResults>> {
-        return kj::NEVER_DONE;
-      }));
+
+  return ioContext.awaitIo(js, kj::mv(callResult),
+      [](jsg::Lock& js, auto result) -> jsg::Value {
+    return jsg::Value(js.v8Isolate, deserializeV8(js, result.getResult().getV8Serialized()));
+  });
 }
 
-kj::Maybe<WorkerRpc::RpcFunction> WorkerRpc::getRpcMethod(jsg::Lock& js, kj::StringPtr name) {
-  // Named intercept is enabled, this means we won't default to legacy behavior.
-  // The return value of the function is a promise that resolves once the remote returns the result
-  // of the RPC call.
-  return RpcFunction([this, methodName=kj::str(name)]
+kj::Maybe<JsRpcCapability::RpcFunction> JsRpcCapability::getRpcMethod(
+    jsg::Lock& js, kj::StringPtr name) {
+  return RpcFunction(JSG_VISITABLE_LAMBDA(
+      (methodName = kj::str(name), self = JSG_THIS),
+      (self),
       (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) -> jsg::Promise<jsg::Value> {
-    auto& ioContext = IoContext::current();
-    // Wait for the RPC to resolve and then process the result.
-    return ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
-        [](jsg::Lock& js, auto result) -> jsg::Value {
-      return jsg::Value(js.v8Isolate, deserializeV8(js, result.getResult().getV8Serialized()));
-    });
-  });
+    // Let's make sure that the returned function wasn't invoked using a different `this` than the
+    // one intended. We don't really _have to_ check this, but if we don't, then we have to forever
+    // support tearing RPC methods off their owning objects and invoking them elsewhere. Enforcing
+    // it might allow future implementation flexibility.
+    //
+    // TODO(cleanup): Ideally, we'd actually unwrap `args.This()` instead of capturing `self` at
+    // all, but there isn't a JSG-bselled way to do that.
```

Comment:
```suggestion
    // all, but there isn't a JSG-based way to do that.
```

Reply:
wat

Reply:
Given that this needs `self` anyway for the call below (`*self->capnpClient`) is this comment entirely accurate? It would seem this needs to hold a strong reference to self either way? Unless you had something else in mind for that bit also?

Reply:
The comment is saying we'd use `unwrap<JsRpcCapability>(args.This())->capnpClient` instead of `self->capnpClient`, if such an `unwrap` function existed.

Reply:
I figured it had to be something along those lines. The comment can be clarified. As currently worded it wouldn't be clear enough to someone else coming into the codebase later trying to figure out how to make the TODO actionable.